File: backend/app/routes/clusters.py
"""
ApeAI — Cluster Routes

Full CRUD for feedback clusters + feedback linking.

Endpoints:
  POST   /clusters/             — Create a new cluster
  GET    /clusters/             — List clusters (with status filter)
  GET    /clusters/{id}         — Get cluster with its feedback items
  PATCH  /clusters/{id}         — Update cluster (title, status, etc.)
  DELETE /clusters/{id}         — Delete a cluster
  POST   /clusters/{id}/feedback — Add feedback items to cluster
  DELETE /clusters/{id}/feedback — Remove feedback from cluster
  GET    /clusters/stats        — Cluster statistics
"""

import logging

from fastapi import APIRouter, HTTPException, Query
from typing import Optional, List

from backend.app.models.storage import (
    ClusterCreate,
    ClusterUpdate,
    ClusterAddFeedback,
    ClusterResponse,
    ClusterListResponse,
)
from backend.app.services.cluster_service import (
    create_cluster,
    get_cluster,
    list_clusters,
    update_cluster,
    delete_cluster,
    add_feedback_to_cluster,
    remove_feedback_from_cluster,
    get_cluster_stats,
)

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/clusters", tags=["Clusters"])


@router.post(
    "/",
    response_model=ClusterResponse,
    summary="Create a new cluster",
)
async def api_create_cluster(data: ClusterCreate):
    """
    Create a new feedback cluster.
    
    Optionally include feedback_ids to link feedback items immediately.
    """
    try:
        cluster = await create_cluster(
            title=data.title,
            summary=data.summary,
            feedback_ids=data.feedback_ids,
            confidence_score=data.confidence_score,
        )
        return _cluster_to_response(cluster)
    except RuntimeError as e:
        raise HTTPException(status_code=503, detail=str(e))


@router.get(
    "/stats",
    summary="Cluster statistics",
)
async def api_cluster_stats():
    """Get cluster count and breakdown by pipeline status."""
    try:
        return await get_cluster_stats()
    except RuntimeError as e:
        raise HTTPException(status_code=503, detail=str(e))


@router.get(
    "/",
    response_model=ClusterListResponse,
    summary="List clusters",
)
async def api_list_clusters(
    status: Optional[str] = Query(None, description="Filter by status"),
    limit: int = Query(50, ge=1, le=200),
    offset: int = Query(0, ge=0),
):
    """List all clusters with optional status filter and pagination."""
    try:
        result = await list_clusters(status=status, limit=limit, offset=offset)
        return ClusterListResponse(
            clusters=[_cluster_to_response(c) for c in result["clusters"]],
            total=result["total"],
        )
    except RuntimeError as e:
        raise HTTPException(status_code=503, detail=str(e))


@router.get(
    "/{cluster_id}",
    response_model=ClusterResponse,
    summary="Get cluster details",
)
async def api_get_cluster(cluster_id: str):
    """Get a cluster including its linked feedback items."""
    try:
        cluster = await get_cluster(cluster_id, include_feedback=True)
        return _cluster_to_response(cluster)
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))
    except RuntimeError as e:
        raise HTTPException(status_code=503, detail=str(e))


@router.patch(
    "/{cluster_id}",
    response_model=ClusterResponse,
    summary="Update cluster",
)
async def api_update_cluster(cluster_id: str, data: ClusterUpdate):
    """Update a cluster's title, summary, status, or confidence score."""
    try:
        cluster = await update_cluster(
            cluster_id=cluster_id,
            title=data.title,
            summary=data.summary,
            status=data.status.value if data.status else None,
            confidence_score=data.confidence_score,
        )
        return _cluster_to_response(cluster)
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))
    except RuntimeError as e:
        raise HTTPException(status_code=503, detail=str(e))


@router.delete(
    "/{cluster_id}",
    summary="Delete cluster",
)
async def api_delete_cluster(cluster_id: str):
    """Delete a cluster and all its feedback links."""
    try:
        await delete_cluster(cluster_id)
        return {"message": f"Cluster {cluster_id} deleted"}
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))
    except RuntimeError as e:
        raise HTTPException(status_code=503, detail=str(e))


@router.post(
    "/{cluster_id}/feedback",
    summary="Add feedback to cluster",
)
async def api_add_feedback(cluster_id: str, data: ClusterAddFeedback):
    """Add feedback items to an existing cluster."""
    try:
        result = await add_feedback_to_cluster(
            cluster_id=cluster_id,
            feedback_ids=data.feedback_ids,
            similarity_scores=data.similarity_scores,
        )
        return result
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))
    except RuntimeError as e:
        raise HTTPException(status_code=503, detail=str(e))


@router.delete(
    "/{cluster_id}/feedback",
    summary="Remove feedback from cluster",
)
async def api_remove_feedback(cluster_id: str, feedback_ids: List[str] = Query(...)):
    """Remove feedback items from a cluster."""
    try:
        result = await remove_feedback_from_cluster(
            cluster_id=cluster_id,
            feedback_ids=feedback_ids,
        )
        return result
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))
    except RuntimeError as e:
        raise HTTPException(status_code=503, detail=str(e))


def _cluster_to_response(cluster: dict) -> ClusterResponse:
    """Convert a raw cluster dict to a ClusterResponse."""
    return ClusterResponse(
        id=cluster.get("id", ""),
        title=cluster.get("title"),
        summary=cluster.get("summary"),
        feedback_count=cluster.get("feedback_count", 0),
        confidence_score=cluster.get("confidence_score", 0.0),
        status=cluster.get("status", "new"),
        created_at=str(cluster.get("created_at", "")),
        updated_at=str(cluster.get("updated_at", "")),
        feedback_items=cluster.get("feedback_items", []),
    )

File: backend/src/routes/clusters.ts
/**
 * ApeAI — Cluster Routes (Layer 2)
 *
 * Full CRUD for feedback clusters + feedback linking.
 * Mirrors Python routes/clusters.py exactly.
 */

import { Router, Request, Response, NextFunction } from 'express';
import { requireAuth } from '../middleware/auth';
import {
  createCluster,
  getCluster,
  listClusters,
  updateCluster,
  deleteCluster,
  addFeedbackToCluster,
  removeFeedbackFromCluster,
  getClusterStats,
} from '../services/clusterService';
import { ClusterCreate, ClusterUpdate, ClusterAddFeedback } from '../types/storage';

export const clustersRouter = Router();

function clusterToResponse(cluster: Record<string, unknown>) {
  return {
    id: cluster['id'] ?? '',
    title: cluster['title'] ?? null,
    summary: cluster['summary'] ?? null,
    feedback_count: cluster['feedback_count'] ?? 0,
    confidence_score: cluster['confidence_score'] ?? 0.0,
    status: cluster['status'] ?? 'new',
    created_at: String(cluster['created_at'] ?? ''),
    updated_at: String(cluster['updated_at'] ?? ''),
    feedback_items: (cluster['feedback_items'] as unknown[]) ?? [],
  };
}

// POST /clusters  (auth required)
clustersRouter.post(
  '/',
  requireAuth,
  async (req: Request, res: Response, next: NextFunction) => {
    const body = req.body as ClusterCreate;
    if (!body.title) {
      res.status(422).json({ detail: "'title' is required" });
      return;
    }
    try {
      const cluster = await createCluster({
        title: body.title,
        summary: body.summary,
        feedbackIds: body.feedback_ids,
        confidenceScore: body.confidence_score,
        userId: req.userId,
      });
      res.status(201).json(clusterToResponse(cluster));
    } catch (err) {
      next(err);
    }
  },
);

// GET /clusters/stats  (auth required) — must come before /:clusterId
clustersRouter.get(
  '/stats',
  requireAuth,
  async (req: Request, res: Response, next: NextFunction) => {
    try {
      const stats = await getClusterStats(req.userId);
      res.json(stats);
    } catch (err) {
      next(err);
    }
  },
);

// GET /clusters  (auth required)
clustersRouter.get(
  '/',
  requireAuth,
  async (req: Request, res: Response, next: NextFunction) => {
    const { status, limit = '50', offset = '0' } = req.query as Record<string, string>;
    try {
      const result = await listClusters({
        status,
        userId: req.userId,
        limit: parseInt(limit, 10),
        offset: parseInt(offset, 10),
      });
      res.json({
        clusters: result.clusters.map(clusterToResponse),
        total: result.total,
      });
    } catch (err) {
      next(err);
    }
  },
);

// GET /clusters/:clusterId
clustersRouter.get(
  '/:clusterId',
  async (req: Request, res: Response, next: NextFunction) => {
    try {
      const cluster = await getCluster(req.params['clusterId']!, true);
      res.json(clusterToResponse(cluster));
    } catch (err) {
      if (err instanceof Error && err.message.includes('not found')) {
        res.status(404).json({ detail: err.message });
        return;
      }
      next(err);
    }
  },
);

// PATCH /clusters/:clusterId
clustersRouter.patch(
  '/:clusterId',
  async (req: Request, res: Response, next: NextFunction) => {
    const body = req.body as ClusterUpdate;
    try {
      const cluster = await updateCluster(req.params['clusterId']!, {
        title: body.title,
        summary: body.summary,
        status: body.status,
        confidenceScore: body.confidence_score,
      });
      res.json(clusterToResponse(cluster));
    } catch (err) {
      if (err instanceof Error && err.message.includes('not found')) {
        res.status(404).json({ detail: err.message });
        return;
      }
      next(err);
    }
  },
);

// DELETE /clusters/:clusterId
clustersRouter.delete(
  '/:clusterId',
  async (req: Request, res: Response, next: NextFunction) => {
    try {
      await deleteCluster(req.params['clusterId']!);
      res.json({ message: `Cluster ${req.params['clusterId']} deleted` });
    } catch (err) {
      if (err instanceof Error && err.message.includes('not found')) {
        res.status(404).json({ detail: err.message });
        return;
      }
      next(err);
    }
  },
);

// POST /clusters/:clusterId/feedback
clustersRouter.post(
  '/:clusterId/feedback',
  async (req: Request, res: Response, next: NextFunction) => {
    const body = req.body as ClusterAddFeedback;
    if (!body.feedback_ids || body.feedback_ids.length === 0) {
      res.status(422).json({ detail: "'feedback_ids' is required" });
      return;
    }
    try {
      const result = await addFeedbackToCluster(
        req.params['clusterId']!,
        body.feedback_ids,
        body.similarity_scores,
      );
      res.json(result);
    } catch (err) {
      if (err instanceof Error && err.message.includes('not found')) {
        res.status(404).json({ detail: err.message });
        return;
      }
      next(err);
    }
  },
);

// DELETE /clusters/:clusterId/feedback
clustersRouter.delete(
  '/:clusterId/feedback',
  async (req: Request, res: Response, next: NextFunction) => {
    const feedbackIds = req.query['feedback_ids'];
    const ids: string[] = Array.isArray(feedbackIds)
      ? (feedbackIds as string[])
      : feedbackIds
      ? [feedbackIds as string]
      : [];

    if (ids.length === 0) {
      res.status(422).json({ detail: "'feedback_ids' query param is required" });
      return;
    }
    try {
      const result = await removeFeedbackFromCluster(req.params['clusterId']!, ids);
      res.json(result);
    } catch (err) {
      if (err instanceof Error && err.message.includes('not found')) {
        res.status(404).json({ detail: err.message });
        return;
      }
      next(err);
    }
  },
);
